stan/math/prim/arr/fun/log_sum_exp.hpp

```diff
@@ -27,21 +29,11 @@ inline double log_sum_exp(const std::vector<double>& x) {
   using std::exp;
   using std::log;
   using std::numeric_limits;
-  double max = -numeric_limits<double>::infinity();
-  for (double xx : x) {
-    if (xx > max) {
-      max = xx;
-    }
-  }
-
-  double sum = 0.0;
-  for (size_t ii = 0; ii < x.size(); ii++) {
-    if (x[ii] != -numeric_limits<double>::infinity()) {
-      sum += exp(x[ii] - max);
-    }
-  }
-
-  return max + log(sum);
+  double max_val = *std::max_element(x.begin(), x.end());
+  double sum = std::accumulate(
+      x.begin(), x.end(), 0.0,
+      [&max_val](auto& acc, auto&& x_i) { return acc + exp(x_i - max_val); });
```

Comment:
Just did this on godbolt, lhs is the code (bottom is current (labled editor 1) and top is the new one (editor 2)  middle is the output from the new stuff and far right is the output from the current stuff. You can highlight certain instructions and it usually pops up a little 'heres what this does'. You can click 'add' in the top right to get a diff view of the two outputs, though it usually looks wonky at O3. You can click and drag any of the tabs for each little block to move stuff. If you right click the highlighted code on the lhs it should have an options to take you to where that line is happening in whichever of the bottom two outputs, though it's not always exact.

I like to look at -O0 to see where stuff is then looking at -O3. About lines 40-60'ish is where the loop and exp calculation happen. The code is super similar, the lambda version removes a compare and a few moves. But those are mostly because we don't do the if statement in there anymore. I can look tmrw at just removing that check there with the old version.

 https://godbolt.org/z/Xe8ev_

godbolt is pretty neat! I learned last night you can also get a real graph of the call graph!

https://godbolt.org/z/cCqIAH

There's a way to make a PR on their repo so we can get Stan up there, would like to find time for that in the next week or so

Reply:
Another cool internet benchmark thing!

http://quick-bench.com/3Wdd56xscm20sShrc0xZx2qgdsE